src/lib/mathlib/math/filter/FOAWDifferentiator.cpp

```diff
@@ -0,0 +1,249 @@
+ *
+ ****************************************************************************/
+
+
+/*
+ * @file    FOAWDifferentiator.cpp
+ * @brief   A class to implement a first order adaptive windowing differentiator
+ * @author Mathieu Bresciani <brescianimathieu@gmail.com>
+ * From:  Discrete-Time Adaptive Windowing for Velocity Estimation
+ * Farrokh Janabi-Sharifi, Vincent Hayward, and Chung-Shin J. Chen
+ */
+
+#include <px4_defines.h>
+#include "FOAWDifferentiator.hpp"
+#include <cmath>
+#include <string.h>
+
+namespace math
+{
+
+FOAWDifferentiator::FOAWDifferentiator(float dt, float noise_level)
+{
+	set_noise_level(noise_level);
+	set_sample_time(dt);
+	reset();
+}
+
+FOAWDifferentiator::~FOAWDifferentiator()
+{
+
+}
+
+void FOAWDifferentiator::set_noise_level(float noise_level)
+{
+	_delta = noise_level;
+}
+
+void FOAWDifferentiator::set_sample_time(float dt)
+{
+	_dt = dt;
+}
+
+uint8_t FOAWDifferentiator::get_last_window_size()
+{
+	return _last_window_size;
+}
+
+float FOAWDifferentiator::get_last_derivative()
+{
+	return _fit_val.a;
+}
+
+void FOAWDifferentiator::reset()
+{
+	memset(&_buffer, 0, sizeof(_buffer));
+	_nb_samples = 0;
+	_last_window_size = 0;
+}
+
+void FOAWDifferentiator::add_sample(float sample)
+{
+	if (_nb_samples <= _max_window_size) {
+		_nb_samples++;
+
+	} else {
+		shift_buffer();
+		_nb_samples = _max_window_size + 1;
+	}
+
+	_buffer[_nb_samples - 1] = sample;
+}
+
+void FOAWDifferentiator::shift_buffer()
+{
+	for (int i = 0; i < (_nb_samples - 1); i++) {
+		_buffer[i] = _buffer[i + 1];
+	}
+}
+
+void FOAWDifferentiator::end_fit_FOAW(uint8_t window_size)
+{
+	float d_amplitude;
+	float d_time;
+	uint8_t last_sample_pos;
+
+	last_sample_pos = _nb_samples - 1;
+
+	d_amplitude = _buffer[last_sample_pos] - _buffer[last_sample_pos - window_size];
+	d_time  = window_size * _dt;
+	_fit_val.a = d_amplitude / d_time;
+	_fit_val.b = _buffer[last_sample_pos];
+}
+
+void FOAWDifferentiator::best_fit_FOAW(uint8_t window_size)
+{
+	float sum1;
+	float sum2;
+	float den;
+	float y_mean;
+	uint8_t last_sample_pos;
+	uint8_t i;
+
+	sum1 = 0.0f;
+	sum2 = 0.0f;
+	den = 0.0f;
+	y_mean = 0.0f;
+	_fit_val.a = 0.0f;
+	_fit_val.b = 0.0f;
+
+	last_sample_pos = _nb_samples - 1;
+
+	// First order least squares fit of all the points inside the window
+	for (i = 0; i <= window_size; i++) {
+		sum1 += _buffer[last_sample_pos - i];
+		sum2 += _buffer[last_sample_pos - i] * i;
+	}
+
+	y_mean = sum1 / (window_size + 1);
+	sum1 *= window_size;
+	sum2 *= 2;
+
+	den = _dt * window_size * (window_size + 1) * (window_size + 2) / 6.0f;
+
+	// Prevents division by zero
+	if (den < 0.0001f && den > -0.0001f) {
+		_fit_val.a = 0.0f;
+
+	} else {
+		_fit_val.a = (sum1 - sum2) / den;
+	}
+
+	_fit_val.b = y_mean + _fit_val.a * window_size * _dt / 2.0f;
+}
+
+// TODO; Add a way to be able to select the method you prefer (End-fit is faster but less accurate)
+// Performs the Best-fit-R algorithm
+float FOAWDifferentiator::fit()
+{
+	uint8_t window_size;
```

Comment:
These can all be `int`'s. It will be faster, because an `int` matches the machines native register size.